### ISSUE TITLE
Give ter-indent a tab or two spaces, and it demands more (inside class/interface)

### ISSUE THREAD

**Issue by ethanova** [buzinas/tslint-eslint-rules#244]
Here's the code:
```
export class ContactCreatedEvent extends MutatingEventBase<ContactPayload, BusinessContext> implements MutatingEvent<ContactPayload, BusinessContext> {
    public static type = 'Contact_Created';
    public static version = 0;
    sequence: number;

    constructor(params: MutatingEventParams<ContactPayload>) {
      super(params);
    }

    apply(context: BusinessContext): BusinessContext {
      context.created = true;
      context.sequence = this.sequence;
      return context;
    }
}
```

Inside es6 classes and interfaces, ter-indent/tslint refuse to indent correctly. Every child line inside the class gives the error `[tslint] Expected indentation of 4 spaces but found 2. (ter-indent)
` and if I concede and give it another 2 spaces it says `[tslint] Expected indentation of 6 spaces but found 4. (ter-indent)` Gives the same issue with tabs. 

It seems to be classes/interfaces with more than one line inside as this doesn't produce the error on the one line inside:
```
export interface ICreateContact extends CQRSCommand<IContactPayload, BusinessContext, OperationalContext> {
  payload: IContactPayload;
}
```
but this does:
```
export interface ICreateContact extends CQRSCommand<IContactPayload, BusinessContext, OperationalContext> {
  payload: IContactPayload;
  test: string;
}
```
Neither of these produce the error:
```
export function test() {
  return null;
}

export function test() {
  console.log('do all the stuff');
  return null;
}
```

Relevant tslint.json
```
"indent": [
    true,
    "spaces",
    2
],
"ter-indent": [
    true,
    2,
    { "SwitchCase": 1 }
],
```

Versions:
- VSCode 1.13.1
- vscode-tslint 0.16.0
- tslint 5.4.3
- typescript 2.4.1
- Node 8.1.3
- Npm 5.1.0
- Windows 10

Let me know what other information or debugging steps I can provide to help!

**Comment by bjspencer** [comment]
I've gotten this as well, but for me this is only occurring when running tslint within vscode using the tslint-language-service, and not when I use the regular vscode-tslint extension.

**Comment by ethanova** [comment]
So upon further investigation it seems to be this line `this.checkNodesIndent(nodesToCheck, indent + indentSize);` inside `blockIndentationCheck` of terIndentRule.js causing the problem. The code treats it as a block and so the nodes inside interface get passed to checkNotesIndent but with a needed indent of 4, which is incorrect. Removing `indentSize` breaks all indentation errors and allows anything. Removing `indent +` seems to fix the issue though. Given that I'm not as familiar with the code base and what that might affect, I didn't want to branch and create a PR until someone more experienced can confirm that's the right move. In the meantime I'll just fix it locally in node_modules.

**Comment by jmlopez-rod** [comment]
@ethanova , Try running the tests and see if your change breaks anything.

**Comment by ethanova** [comment]
@jmlopez-rod when I run `npm test` right after pulling down master and npm installing I get this error:
```
> tslint-eslint-rules@4.1.1 test D:\projects\tslint-eslint-rules
> gulp test && gulp self-lint

[14:04:18] Using gulpfile D:\projects\tslint-eslint-rules\gulpfile.js
[14:04:18] Starting 'lint'...
Warning: The 'no-use-before-declare' rule requires type infomation.
D:/projects/tslint-eslint-rules/src/test/rules/ruleTester.ts
ERROR: 264:15  no-shadowed-variable  Shadowed name: 'configFile'

[14:04:20] 'lint' errored after 1.89 s
[14:04:20] Error in plugin 'gulp-tslint'
Message:
    Failed to lint: D:/projects/tslint-eslint-rules/src/test/rules/ruleTester.ts [264, 15]: Shadowed name: 'configFile'.
npm ERR! Test failed.  See above for more details.
```

**Comment by ethanova** [comment]
After fiixing that error and running the tests, they all work. But after making my change, tests fail as shown in the attached.
[postIndentChangeTests2.txt](https://github.com/buzinas/tslint-eslint-rules/files/1149353/postIndentChangeTests2.txt)

**Comment by Kovensky** [comment]
I hit this earlier today when I tried to `extend` a config file from a different folder that would hit a different `node_modules` folder (bad idea, I know; "temporary" hack).

This hints that the problem might be caused by `ter-indent` loading a different version of `typescript` than the version of `typescript` being used by `tslint` (or, in vscode's case, the builtin `typescript`). Try setting the `typescript.tsdk` to point to your project's `typescript`.

**Comment by ethanova** [comment]
hey @Kovensky thanks for the suggestion. I tried setting `"typescript.tsdk": "node_modules/typescript/lib"` in my settings.json at the project root, installing typescript there, and restarting which didn't make the error go away.

**Comment by jmlopez-rod** [comment]
@ethanova Check in `node_modules/tslint` to see if you see the subfolder `node_modules/typescript`. I have come across cases in which some of dependencies (typedoc for instance) depends on very specific version of typescript and it gets quite hard to figure out what version of typescript a dependency is using.

**Comment by ethanova** [comment]
![image](https://user-images.githubusercontent.com/1507541/29832625-7128e0a2-8cae-11e7-933b-88b5e3b3ec52.png)
I don't see that unfortunately. I noticed yesterday when I was working with my PATH variable that apparently some version of Visual Studio installed Typescript 1.0 and put that on my PATH. It could be that's what is causing it? I don't have time at the moment to go and delete it from my PATH and check, Unfortunately my PATH is over 2047 characters so I can't manage it from the GUI and haven't had time to clean it up.

**Comment by kasperoo** [comment]
Seeing the same problem. Just after we upgraded tslint from 5.5.0 to 5.6.0

**Comment by nfour** [comment]
Appears to only occur on typescript@2.5 for me.

**Comment by jmlopez-rod** [comment]
Is this still an issue? I recall trying this out and I did not encounter the problem. I opened a branch to test this out:

```
ruleTester.addTestGroupWithConfig(
  'issue-244',
  'indentation inside classes',
  [
    2,
    { SwitchCase: 1 }
  ],
  [
  {
    code: dedent`
    export function test() {
      return null;
    }

    export function test() {
      console.log('do all the stuff');
      return null;
    }
    `
  },
  {
    code: dedent`
    export interface ICreateContact extends CQRSCommand<IContactPayload, BusinessContext, OperationalContext> {
      payload: IContactPayload;
      test: string;
    }
    `
  },
  {
    code: dedent`
    export interface ICreateContact extends CQRSCommand<IContactPayload, BusinessContext, OperationalContext> {
      payload: IContactPayload;
    }`
  },
  {
    code: dedent`
    export class ContactCreatedEvent extends MutatingEventBase<ContactPayload, BusinessContext> implements MutatingEvent<ContactPayload, BusinessContext> {
        public static type = 'Contact_Created';
        public static version = 0;
        sequence: number;

        constructor(params: MutatingEventParams<ContactPayload>) {
          super(params);
        }

        apply(context: BusinessContext): BusinessContext {
          context.created = true;
          context.sequence = this.sequence;
          return context;
        }
    }`
    }
]);
```

After running the test I get this:

```
$ gulp test --single ter-indent:issue-244
[14:08:36] Using gulpfile ~/Workspace/tslint-eslint-rules/gulpfile.js
[14:08:36] Starting 'lint'...
Warning: The 'no-use-before-declare' rule requires type information.
[14:08:36] Finished 'lint' after 407 ms
[14:08:36] Starting 'build'...
[14:08:38] Finished 'build' after 1.73 s
[14:08:38] Starting 'test'...


  ter-indent
    1) issue-244 - indentation inside classes


  0 passing (61ms)
  1 failing

  1) ter-indent issue-244 - indentation inside classes:
     AssertionError: Error mismatch in issue-244-3.ts:

      0| 
      1| export class ContactCreatedEvent extends MutatingEventBase<ContactPayload, BusinessContext> implements MutatingEvent<ContactPayload, BusinessContext> {
      2|     public static type = 'Contact_Created';
      3|     public static version = 0;
      4|     sequence: number;
      5| 
      6|     constructor(params: MutatingEventParams<ContactPayload>) {
      7|       super(params);
      8|     }
      9| 
     10|     apply(context: BusinessContext): BusinessContext {
     11|       context.created = true;
     12|       context.sequence = this.sequence;
     13|       return context;
     14|     }
     15| }

     

     Found:
       [issue-244-3.ts@{[2:0|153] -> [2:4|157]}] ter-indent: Expected indentation of 2 spaces but found 4.
       [issue-244-3.ts@{[3:0|197] -> [3:4|201]}] ter-indent: Expected indentation of 2 spaces but found 4.
       [issue-244-3.ts@{[4:0|228] -> [4:4|232]}] ter-indent: Expected indentation of 2 spaces but found 4.
       [issue-244-3.ts@{[6:0|251] -> [6:4|255]}] ter-indent: Expected indentation of 2 spaces but found 4.
       [issue-244-3.ts@{[10:0|342] -> [10:4|346]}] ter-indent: Expected indentation of 2 spaces but found 4.

      at Test.compareErrors (dist/test/rules/ruleTester.js:97:16)
      at Test.runTest (dist/test/rules/ruleTester.js:72:14)
      at dist/test/rules/ruleTester.js:211:38
      at Array.forEach (native)
      at Context.<anonymous> (dist/test/rules/ruleTester.js:209:37)



[14:08:38] 'test' errored after 537 ms
[14:08:38] Error in plugin 'gulp-spawn-mocha'
Mocha exited with code 1
```

Those are valid errors, we are specifying the use of two spaces per indent. I'll close this for now but feel free to reopen if the problem can be reproduced.

**Comment by lafiosca** [comment]
I am encountering this same error, but in my case it appears to be caused by an interaction between globally-installed and locally-installed packages. Specifically, I was under a project directory generated by Angular's `ng new` command, which had installed tslint@5.3.2, but the tslint I was running was the globally-installed tslint@5.7.0. Some of the relevant dependencies also had differing versions.

I believe the solution is either (1) to only run the locally-installed tslint from `node_modules/.bin` within the project, or (2) to ensure that all of the local and global versions match. Because I experienced this via vim's Syntastic plugin, which was executing the globally-installed tslint, so it was easier for me to take approach 2 by running `npm i tslint@latest tsutils@latest tslint-config-airbnb-base@latest tslint-eslint-rules@latest typescript@latest` in the Angular project. An alternative for use with approach 1 would have been to install a vim plugin to allow Syntastic to find and execute the locally-installed tslint instead. (I think I took the lazier approach.)

I am sharing this just in case others like me stumble across this thread in search of help because it was difficult at first for me to find information or decipher what was happening. Thank you to the folks above who shared enlightening tidbits on this.
